// naar-music-bot/src/commands/queue.js
const { MessageEmbed, splitMessage } = require("discord.js");

module.exports = {
    name: "queue",
    aliases: ["qu"],
    cooldown: 10,
    description: `Shows you the songs queue!.`,
    run: async(client, message, args) => {
        if (client.queue_command == true) {
            if (!message.guild) return;
            const queue = message.client.queue.get(message.guild.id);
            if (!queue) return message.channel.send(
                new MessageEmbed()
                .setTitle("Thare is nothing in the music queue!")
                .setColor("RED")).catch(err => {
                throw new TypeError(err)
            });
            console.log(queue.songs);
            let description = "";
            for (let i = 0; i < queue.songs.length; i++) {
                description += `**${i}.** [${queue.songs[i].title.substring(0, 40)}](${queue.songs[i].url}) | \`${queue.songs[i].duration}\`\n`
            }
            let queueEmbed = new MessageEmbed()
                .setTitle("Music Queue")
                .setDescription(description)
                .setColor("#c219d8");
            const splitDescription = splitMessage(description, {
                maxLength: 2048,
                char: "\n",
                prepend: "",
                append: ""
            });
            splitDescription.forEach(async(m) => {
                queueEmbed.setDescription(m);
                message.react("✅")
                message.channel.send(queueEmbed);
            });
        } else if (client.help_command == false) {
            message.channel.send(
                new MessageEmbed()
                .setTitle(`\`${module.exports.name}\` Has Been Disabled From The Music System`)
            )
        } else throw new TypeError(`❌ | "${module.exports.name}_command" value must be true or false`)
    }
};
